uaclient/cli.py

```diff
@@ -429,7 +429,16 @@ def action_status(args, cfg):
             status["expires"] = str(status["expires"])
         print(json.dumps(status))
     else:
-        print(ua_status.format_tabular(cfg.status()))
+        output = ua_status.format_tabular(cfg.status())
+        # Replace our Unicode dash with an ASCII dash if we aren't going to be
+        # writing to a utf-8 output; see
+        # https://github.com/CanonicalLtd/ubuntu-advantage-client/issues/859
+        if (
+            sys.stdout.encoding is None
+            or "UTF-8" not in sys.stdout.encoding.upper()
+        ):
+            output = output.replace("\u2014", "-")
```

Comment:
This may fall into the category of your PR description:  "This is specifically targeted to fix #859, there may be more general changes to follow."   

Food for thought: if we don't think that "\u2014" is an appropriate character sequence to show people in the console status output, shouldn't  we also replace this string in the status.json output too in general so that is is more human-friendly?

Reply:
> This may fall into the category of your PR description: "This is specifically targeted to fix #859, there may be more general changes to follow."

Yes, though given the below I'm not sure we actually need more general changes.

> Food for thought: if we don't think that "\u2014" is an appropriate character sequence to show people in the console status output, shouldn't we also replace this string in the status.json output too in general so that it is more human-friendly?

The status value stored in `status.json` and (therefore) emitted when `--output json` is passed to `ua status` is _not_  a string containing the single Unicode character `\u2014` (i.e. —), it's the string "\u2014". It's already been encoded, specifically because `ensure_ascii` defaults to True for [`json.dump(s)`](https://docs.python.org/3/library/json.html#json.dump).

>  so that it is more human-friendly?

This is JSON output, it doesn't need to be human-friendly. (And, to be clear, we aren't avoiding emitting it in those other cases to be "human-friendly" except insofar as breaking someone's terminal isn't very friendly to that human.)

Reply:
I didn't realize as well that when rendering that json output in my terminal with utf-8 encoding it actually is converted back to the appropriate em-dash character. For instance: `ua status --format json` output will still contain the "\u2014" unicode character representation of the character but piping that to another tool such as jq or json_pp will result in  rendering the em-dash character "—". It makes potential script matches a bit more complex, but it's just an opinion and not a blocker to this branch especially because the status value of "—" is of dubious meaning anyway.  It's probably more likely people would be searching for strings enabled|disabled|n/a  as they aren't vague.


But here's an example of the concern, scripts that directly process `ua status --format json` ouput may incorrectly parse the status column if they are not aware of the differences. Yay unicode encoding issues.

root@dev-t:~# ua status --format json | grep —
root@dev-t:~# ua status --format json | jq "." | grep —
      "status": "—",
root@dev-t:~# ua status --format json | json_pp | grep —
Wide character in print at /usr/bin/json_pp line 81, <STDIN> chunk 1.
         "status" : "—",
root@dev-t:~#

Reply:
This seems very clearly to be a distinct issue, as it's about the external usability of our experimental, internal-only JSON representation. Can you file one, please?

> not a blocker to this branch

So can I get an Approve?

Reply:
indeed.